### PR TITLE
Add ability to provision infrastructure during Azure Container Instance push work pool creation

### PR DESCRIPTION
**Pull request by serinamarie**
```bash
❯ prefect work-pool create my-work-pool --provision-infra
16:28:26.763 | DEBUG   | prefect.profiles - Using profile 'cloud'
16:28:26.857 | DEBUG   | prefect.client - Connecting to API at https://api.prefect.cloud/api/accounts/abc/workspaces/xyz/
? What type of work pool infrastructure would you like to use? [Use arrows to move; enter to select]
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃    ┃ Infrastructure Type                  ┃ Description                                                                                                                                                         ┃
┡━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│    │ Prefect Agent                        │ Execute flow runs on heterogenous infrastructure using infrastructure blocks.                                                                                       │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Local Subprocess                     │ Execute flow runs as subprocesses on a worker. Works well for local execution when first getting started.                                                           │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ AWS Elastic Container Service        │ Execute flow runs within containers on AWS ECS. Works with EC2 and Fargate clusters. Requires an AWS account.                                                       │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Azure Container Instances            │ Execute flow runs within containers on Azure's Container Instances service. Requires an Azure account.                                                              │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Docker                               │ Execute flow runs within Docker containers. Works well for managing flow execution environments via Docker images. Requires access to a running Docker daemon.      │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Google Cloud Run                     │ Execute flow runs within containers on Google Cloud Run. Requires a Google Cloud Platform account.                                                                  │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Cloud Run Worker V2                  │ A worker which runs flow runs on Google Cloud Run (API v2).                                                                                                         │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Google Vertex AI                     │ Execute flow runs within containers on Google Vertex AI. Requires a Google Cloud Platform account.                                                                  │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Kubernetes                           │ Execute flow runs within jobs scheduled on a Kubernetes cluster. Requires a Kubernetes cluster.                                                                     │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ Google Cloud Run - Push              │ Execute flow runs within containers on Google Cloud Run. Requires a Google Cloud Platform account. Flow runs are pushed directly to your environment, without the   │
│    │                                      │ need for a Prefect worker.                                                                                                                                          │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│    │ AWS Elastic Container Service - Push │ Execute flow runs within containers on AWS ECS. Works with existing ECS clusters and serverless execution via AWS Fargate. Requires an AWS account. Flow runs are   │
│    │                                      │ pushed directly to your environment, without the need for a Prefect worker.                                                                                         │
├────┼──────────────────────────────────────┼─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ >  │ Azure Container Instances - Push     │ Execute flow runs within containers on Azure's Container Instances service. Requires an Azure account. Flow runs are pushed directly to your environment, without   │
│    │                                      │ the need for a Prefect worker.                                                                                                                                      │
└────┴──────────────────────────────────────┴─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
16:28:30.319 | DEBUG   | prefect.client - Connecting to API at https://api.prefect.cloud/api/accounts/abc/workspaces/xyz/
? Please select which Azure subscription to use: [Use arrows to move; enter to select]
┏━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃    ┃ Name                 ┃ Subscription ID                      ┃
┡━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│    │ Engineering          │ 123 │
│ >  │ Azure subscription 1 │ 234 │
│    │ Azure subscription 1 │ 345 │
│    │ subscription2        │ 456 │
│    │ subscription2        │ 567 │
└────┴──────────────────────┴──────────────────────────────────────┘
╭───────────────────────────────────────────────────────────────────────────────────────╮
│ Provisioning infrastructure for your work pool s will require:                        │
│                                                                                       │
│     Updates in subscription Azure subscription 1                                      │
│                                                                                       │
│         - Create a resource group in location eastus                                  │
│         - Create an app registration in Azure AD                                      │
│         - Create a service principal for app registration                             │
│         - Generate a secret for app registration                                      │
│         - Assign Contributor role to service account                                  │
│         - Create Azure Container Instance                                             │
│                                                                                       │
│     Updates in Prefect workspace                                                      │
│                                                                                       │
│         - Create Azure Container Instance credentials block aci-push-pool-credentials │
│                                                                                       │
╰───────────────────────────────────────────────────────────────────────────────────────╯
Proceed with infrastructure provisioning? [y/n]: y
Creating resource group
Provisioning infrastructure... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% -:--:--Resource group 'prefect-aci-push-pool-rg' created in location 'eastus'
Creating app registration
Provisioning infrastructure... ━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━  20% -:--:--App registration 'prefect-aci-push-pool-app' created successfully
Generating secret for app registration
Provisioning infrastructure... ━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━  40% 0:00:06Secret generated for app registration with client ID '1111cb05-3b9e-4d1b-92cc-d897b2d4f337'
ACI credentials block 'y-push-pool-credentials' created
Assigning Contributor role to service account...
Provisioning infrastructure... ━━━━━━━━━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━  60% 0:00:06Contributor role assigned to service principal with object ID '4be6fed7-ecbd-4941-9da8-16af068f2a45'
Creating Azure Container Instance
Provisioning infrastructure... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╺━━━━━━━  80% 0:00:04Container instance 'prefect-acipool-container' created successfully
Creating Azure Container Instance credentials block
Provisioning infrastructure... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
Created work pool 'my-work-pool'!
```

If infrastructure already exists:
```bash
Proceed with infrastructure provisioning? [y/n]: y
Creating resource group
Resource group 'prefect-aci-push-pool-rg' already exists in location 'eastus'.
Creating app registration
App registration 'prefect-aci-push-pool-app' already exists.
Generating secret for app registration
Provisioning infrastructure... ━━━━━━━━━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━  40% 0:00:02Secret generated for app registration with client ID '1111cb05-3b9e-4d1b-92cc-d897b2d4f337'
ACI credentials block 'bb-push-pool-credentials' created
Assigning Contributor role to service account...
Service principal with object ID '4be6fed7-ecbd-4941-9da8-16af068f2a45' already has the 'Contributor' role assigned in 
'/subscriptions/2d38d768-2bc9-46b1-af48-5568eeea417c/resourceGroups/prefect-aci-push-pool-rg'
Creating Azure Container Instance
Container instance 'prefect-aci-push-pool-container' already exists.
Creating Azure Container Instance credentials block
Provisioning infrastructure... ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00
Infrastructure successfully provisioned!
Created work pool 'my-work-pool'!
``` 

### Checklist

- [ ] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [ ] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`, `docs`.

### CODE REVIEW

File: src/prefect/infrastructure/provisioners/container_instance.py
Comment (iotmani) [inline src/prefect/infrastructure/provisioners/container_instance.py:1]: IMO it might be clearer to call it something like `azure_container_instances.py`?

Comment (serinamarie) [reply src/prefect/infrastructure/provisioners/container_instance.py:1]: I agree; I was matching with the naming convention of the other provisioner file `cloud_run.py`

File: src/prefect/infrastructure/provisioners/container_instance.py [hunk 1]
```diff
@@ -0,0 +1,765 @@
...
+            )
+
+            self._console.print(
+                (
+                    f"ACI credentials block '{credentials_block_name}' created in"
+                    " Prefect Cloud"
+                ),
+                style="green",
+            )
+            return block_doc.id
```
Comment (desertaxle) [inline src/prefect/infrastructure/provisioners/container_instance.py:613]: Note to future us: We're doing this in at least 4 places now, so it's probably time to create a utility.

File: src/prefect/infrastructure/provisioners/container_instance.py [hunk 2]
```diff
@@ -0,0 +1,765 @@
...
+                )
+                progress.advance(task)
+            else:
+                progress.console.print("ACI credentials block already exists")
+                block_doc = await client.read_block_document_by_name(
+                    name=f"{work_pool_name}-push-pool-credentials",
+                    block_type_slug="azure-container-instance-credentials",
+                )
+                block_doc_id = block_doc.id
+                progress.advance(task)
```
Comment (desertaxle) [inline src/prefect/infrastructure/provisioners/container_instance.py:741]: We'll want to advance by 2 here to ensure the amount advanced is asymmetric between the two branches.
```suggestion
                progress.advance(task, advance=2)
```

File: src/prefect/infrastructure/provisioners/container_instance.py [hunk 3]
```diff
@@ -0,0 +1,765 @@
...
+                    app_registration = json.loads(app_registration)
+                return app_registration["appId"]
+
+        elif result == "exists":
+            self._console.print(
+                f"App registration '{self.APP_REGISTRATION_NAME}' already exists.",
+                style="yellow",
+            )
+        elif result == "error":
+            raise Exception("Error creating the app registration.")
```
Comment (desertaxle) [inline src/prefect/infrastructure/provisioners/container_instance.py:365]: Looks like there are two different modes of operation here:

1. You run a separate command to check if something exists
2. You let `AzureCLI` let you know if something already existed

This doesn't need to be addressed in this PR, but having a consistent sequence for each thing getting created would be nice. My guess is the different approaches are probably due to the idiosyncrasies of the `az` CLI, but it'd be nice to find a way to hide that implementation detail.

File: tests/infrastructure/provisioners/test_container_instance.py [hunk 4]
```diff
@@ -0,0 +1,654 @@
+from subprocess import CalledProcessError
+from unittest.mock import AsyncMock, MagicMock, call
+
+import pytest
+
+from prefect.infrastructure.provisioners.container_instance import (
+    ContainerInstancePushProvisioner,
+)
+
+
```
Comment (desertaxle) [inline tests/infrastructure/provisioners/test_container_instance.py:10]: Can you add tests for these cases:
1. Call `.provision`, `az` not installed
2. Call `.provision`, no preexisting credentials block
3. Call `.provision`, existing credentials block

In the last two cases, we should check to ensure the base job template is configured with the expected resources.

**Review by desertaxle — COMMENTED** [review]
The implementation is looking good! I mainly just want to make sure we have test coverage for `.provision`.

File: src/prefect/infrastructure/provisioners/container_instance.py [hunk 5]
```diff
@@ -0,0 +1,765 @@
...
+        """
+        try:
+            command = (
+                'az account list-locations --query "[?isDefault].name" --output tsv'
+            )
+            _, output = await self.azure_cli.run_command(command)
+            self._location = output if output else self.DEFAULT_LOCATION
+        except subprocess.CalledProcessError as e:
+            self._location = self.DEFAULT_LOCATION
+            raise RuntimeError("Failed to get default location.") from e
```
[the remark this answers is not in the document]
Comment (serinamarie) [reply src/prefect/infrastructure/provisioners/container_instance.py:179]: I think this was just from refactoring. Will remove

File: src/prefect/infrastructure/provisioners/container_instance.py (hunk 2 above)
Comment (serinamarie) [reply src/prefect/infrastructure/provisioners/container_instance.py:753]: Oops, I forgot to push this:
https://github.com/PrefectHQ/prefect/pull/11275/files#diff-cfa849f982ca38f020a3d71caefe80bd5b7e78dbb3a2186875853964e54e75c3R711-R718

do you prefer using `advance` over that?

File: tests/infrastructure/provisioners/test_container_instance.py (hunk 4 above)
Comment (serinamarie) [reply tests/infrastructure/provisioners/test_container_instance.py:10]: Will do!

File: src/prefect/infrastructure/provisioners/container_instance.py (hunk 2 above)
Comment (desertaxle) [reply src/prefect/infrastructure/provisioners/container_instance.py:753]: Yup that will work!

File: tests/infrastructure/provisioners/test_container_instance.py (hunk 4 above)
Comment (serinamarie) [reply tests/infrastructure/provisioners/test_container_instance.py:22]: I've applied those changed in subsequent commits :)

File: src/prefect/infrastructure/provisioners/container_instance.py (hunk 3 above)
Comment (serinamarie) [reply src/prefect/infrastructure/provisioners/container_instance.py]: That makes sense. I had started in one direction (mode 2) but ended up converting to mode 1 as it felt more thorough. This should be consistent across the board now in [0694586](https://github.com/PrefectHQ/prefect/pull/11275/commits/0694586696b0a3c81c23fa6b477234b743a8d730)

**Pull request merged**